kubespawner/spawner.py

```diff
@@ -2577,6 +2593,37 @@ async def _make_delete_pod_request(
             else:
                 raise
 
+    async def _make_delete_pvc_request(self, pvc_name, request_timeout):
+        """
+        Make an HTTP request to delete the given PVC
+
+        Designed to be used with exponential_backoff, so returns
+        True / False on success / failure
+        """
+        self.log.info("Deleting pvc %s", pvc_name)
+        try:
+            await gen.with_timeout(
+                timedelta(seconds=request_timeout),
+                self.asynchronize(
+                    self.api.delete_namespaced_persistent_volume_claim,
+                    name=pvc_name,
+                    namespace=self.namespace,
+                ),
+            )
+            return True
+        except gen.TimeoutError:
+            return False
+        except ApiException as e:
+            if e.status == 404:
+                self.log.warning(
+                    "No pvc %s to delete. Assuming already deleted.",
+                    pvc_name,
+                )
+                # If there isn't a PVC to delete, that's ok too!
```

Comment:
I've got a question about this and the warning a couple of lines above. In our deployment we're currently using kubespawner 0.15 and jupyterhub 1.3. I'm looking to upgrade to z2jh 1.2.0, jupyterhub 1.5.0 and kubespawner 1.1.2. We use a single PVC that is backed by cloud object storage so all of the singleuser-server pods share the same PVC. We leave the default of `storage_pvc_ensure=False` which from looking at the code means kubespawner won't actually create a PVC named `self.pvc_name` in `_start`. So shouldn't `delete_forever` also be checking `storage_pvc_ensure`? Maybe not if it's assumed that someone could pre-create the PVC using the same `pvc_name_template`, and in that case we just try to delete the PVC and if it doesn't exist we get a 404 and log the warning.

I guess in our case the point is I don't want to see the warnings when users get culled (which happens after 5 days of inactivity for us) so to avoid that we just need to set `delete_pvc=False`.

Reply:
I extracted this to https://github.com/jupyterhub/kubespawner/issues/543 to avoid loosing track of this comment!